### DISCUSSION TITLE
Features

### DISCUSSION THREAD

**Discussion by USLTD** [directvt/vtm#174] [Ideas]
# Applications and APP Engine

I have seen few applications in source and also on demo and I have a question. Are applications built-in only (meaning that only applications which are compiled during **vtm** compilation are present)? And if yes are you planning to add this following feature:

> Create and link some kind of directory for **vtm**. Like **.vtm** which will contain applications (like **.vtm/apps/**), settings (possibly **JSON** or **INI** format) and possibly extensions
> During running **vtm** will always watch **.vtm/apps/** directory and in case of changes it will add/remove it in apps list and ...
> -  In case of removal it will display error dialog like *"{application_name} got removed!"* and close application
> - In case of modification of existing it will display choice box with message *"{application_name} got modified. Do you want to reload?"* and do according user's choice (close and reopen on **Yes** or continue with loaded application if **No**)
> - In case of addition just add new application and it's name in app list

**Comment by o-sdn-o** [comment]
Thanks for your suggestion, for some application types this will probably be necessary. But I try to reduce the number of built-in applications to zero, and here's why.

On the one hand, it is convenient to have built-in applications, this simplifies their interaction with the vtm environment, but there are significant drawbacks:
  - applications occupy vtm's UI thread
  - applications allocate the memory of the main vtm process, the total memory consumption increases due to fragmentation, and this is critical for long-running processes such as the vtm environment

At the moment, demo applications are built into vtm and compiled with it to simplify the desktopio framework demo. In the future, all applications will be independent applications (standalone TUI executable) capable of running in any terminal, including vtm itself. Only the terminal emulator (required to run external TUIs) and the tiling window manager will remain the built-in vtm applications.

The application list (main menu items) will be automatically synchronized with the vtm settings file (~/.config/vtm/settings.ini). In the settings window it will be possible to make all the necessary changes using the UI, at the moment there is only the frame rate configuration.

**Comment by USLTD** [comment]
Alright. Also will you also build web- and files browser? Especially save/open dialogs?

**Comment by o-sdn-o** [comment]
Of course, external applications need an API to support modal windows (e.g. save/open dialogs). Most likely these will be asynchronous requests using vt-sequences. vtm will draw standard modal windows on its own, while non-standard ones will require some kind of non-standard solution.

**Comment by o-sdn-o** [comment]
Something like the following

```
ESC _msg_box_cmd_; title; body; button_1; ... ; button_n  ST
```

I'll also be able to let external applications create new windows and write to them via stdout.

**Comment by USLTD** [comment]
Also are you planning to add wrappers and bindings? Like how **notcurses** has <kbd>Python</kbd> wrappers and bindings?

**Comment by USLTD** [comment]
Also when you will finish engine and API, in another words complete **vtm**'s API and features, I am planning to translate those <kbd>C++</kbd> code into another languages. For example into <kbd>C#</kbd> or <kbd>Python</kbd>. <kbd>Python</kbd>'s version would be slower but in exchange it would be easier to create own widgets, apps and more.

**Comment by o-sdn-o** [comment]
> Also are you planning to add wrappers and bindings? Like how notcurses has Python wrappers and bindings?

Yes, it's definitely needed.

**Reply by USLTD** [reply]
I think it's time for me to learn <kbd>C++</kbd> 😁

**Comment by USLTD** [comment]
@o-sdn-o 
I have a suggestion. Why don't make **Windows** style  start menu and taskbar as a alternative style? Current **vtm** start menu is good but I think it will be more familiar and better with **Windows** style. Although this would be default for **Windows**'s version, although user will be able to change launcher into **vtm** (current), **MacOS** or **Linux** styles.

I even created mini **Windows** logo using **ANSI**
![IMG_20220215_163736](https://user-images.githubusercontent.com/57063715/154064769-306441f1-3c2c-442b-b641-ea3c43589e49.jpg)

Here is source: `\x1b[48;5;202;38;5;39m▄\x1b[48;5;106;38;5;214m▄\x1b[49;38;5;33m Microsoft\x1b[m`.

Or alternative `\x1b[48;2;0;164;239m\x1b[38;2;242;80;34m▀\x1b[48;2;255;185;0m\x1b[38;2;127;186;0m▀` (which looks better)
![IMG_20220215_170147](https://user-images.githubusercontent.com/57063715/154067440-534aa1b6-ee75-463c-ab53-ca454a849818.jpg)

**Reply by USLTD** [reply]
I even have **Windows 11**'s logo if you want but it's currently not good.

**Reply by USLTD** [reply]
![IMG_20220215_170506](https://user-images.githubusercontent.com/57063715/154067765-800c5f0a-5485-414d-a043-49e216da194e.jpg)

**Reply by USLTD** [reply]
Still figuring out how to make it look better

**Reply by o-sdn-o** [reply]
The start menu is a built-in application. When the Desktopio API and additional tools become available, it will be easy to create any start menu you like. Now the structure of such an application has to be coded at a low level, and it is unproductive to modify it without additional tools such as UI Builder.

**Reply by o-sdn-o** [reply]
It will be cool to make all sorts of menus stylized for various desktop operating systems 😊
